src/liballoc/vec.rs

```diff
@@ -603,6 +603,92 @@ impl<T> Vec<T> {
+    ///
+    /// # Example
+    ///
+    /// This API is useful especially when using `Vec` as a
+    /// temporary storage for data with short lifetimes.
+    /// By recycling the allocation, the `Vec` is able to safely
+    /// outlive the lifetime of the type that was stored in it.
+    /// ```
+    /// # #![feature(recycle_vec)]
+    /// # use std::error::Error;
+    /// #
+    /// # struct Stream(bool);
+    /// #
+    /// # impl Stream {
+    /// #     fn new() -> Self {
+    /// #         Stream(false)
+    /// #     }
+    /// #
+    /// #     fn next(&mut self) -> Option<&[u8]> {
+    /// #         if self.0 {
+    /// #             None
+    /// #         } else {
+    /// #             self.0 = true;
+    /// #             Some(&b"foo"[..])
+    /// #         }
+    /// #     }
+    /// # }
+    /// #
+    /// # fn process(input: &[Object<'_>]) -> Result<(), Box<dyn Error>> {
+    /// #     for obj in input {
+    /// #         let _ = obj.reference;
+    /// #     }
+    /// #     Ok(())
+    /// # }
+    /// #
+    /// # struct Object<'a> {
+    /// #     reference: &'a [u8],
+    /// # }
+    /// #
+    /// # fn deserialize<'a>(input: &'a [u8], output: &mut Vec<Object<'a>>) -> Result<(), Box<dyn Error>> {
+    /// #     output.push(Object { reference: input });
+    /// #     Ok(())
+    /// # }
+    /// #
+    /// # fn main() -> Result<(), Box<dyn Error>> {
+    /// # let mut stream = Stream::new();
+    /// let mut objects: Vec<Object<'static>> = Vec::new();    // Any lifetime goes here
+    ///
+    /// while let Some(byte_chunk) = stream.next() {           // `byte_chunk` lifetime starts
+    ///     let mut temp: Vec<Object<'_>> = objects.recycle(); // `temp` lifetime starts
+    ///
+    ///     // Zero-copy parsing; deserialized `Object`s have references to `byte_chunk`
+    ///     deserialize(byte_chunk, &mut temp)?;
+    ///     process(&temp)?;
+    ///
+    ///     objects = temp.recycle();                          // `temp` lifetime ends
+    /// }                                                      // `byte_chunk` lifetime ends
+    /// # Ok(())
+    /// # }
+    /// ```
+    ///
+    /// # Panics
+    /// Panics if the size or alignment of the source and target types don't match.
+    ///
+    /// # Note about stabilization
+    /// The size and alignment contract is enforceable at compile-time,
+    /// so we will want to wait until compile-time asserts become stable and
+    /// modify this API to cause a compile error instead of panicking
+    /// before stabilizing it.
+    #[unstable(feature = "recycle_vec", reason = "new API", issue = "0")]
+    pub fn recycle<U>(mut self) -> Vec<U> {
+        self.truncate(0);
+        // TODO make these const asserts once it becomes possible
+        assert_eq!(core::mem::size_of::<T>(), core::mem::size_of::<U>());
+        assert_eq!(core::mem::align_of::<T>(), core::mem::align_of::<U>());
+        let capacity = self.capacity();
```

Comment:
Please use https://doc.rust-lang.org/nightly/std/vec/struct.Vec.html#method.into_raw_parts instead.

Reply:
Thanks for pointing this out, I didn't know about this API. I also change `truncate` to `clear` as it states the intent better.